File: src/ClientModel/ClientModels.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

using System.Collections;
using System.Collections.Generic;

namespace AutoRest.Java.Model
{
    /// <summary>
    /// The collection of all client models stored for inheritance lookup.
    /// </summary>
    public class ClientModels : IEnumerable<ClientModel>
    {
        public static readonly ClientModels Instance = new ClientModels();

        private ClientModels() { }

        private readonly IDictionary<string, ClientModel> nameMap = new Dictionary<string, ClientModel>();
        private readonly IDictionary<string, List<ClientModel>> derivedTypesMap = new Dictionary<string, List<ClientModel>>();

        public ClientModel GetModel(string modelName)
        {
            return nameMap.ContainsKey(modelName) ? nameMap[modelName] : null;
        }

        public void AddModel(ClientModel model)
        {
            nameMap[model.Name] = model;

            ClientModel parentModel = model.ParentModel;
            if (parentModel != null)
            {
                List<ClientModel> derivedTypesList = GetDerivedTypeList(parentModel.Name);
                derivedTypesList.Add(model);
            }
        }

        public IEnumerable<ClientModel> GetDerivedTypes(string parentModelName)
            => GetDerivedTypeList(parentModelName);

        private List<ClientModel> GetDerivedTypeList(string parentModelName)
        {
            if (!derivedTypesMap.ContainsKey(parentModelName))
            {
                derivedTypesMap.Add(parentModelName, new List<ClientModel>());
            }
            return derivedTypesMap[parentModelName];
        }

        public IEnumerator<ClientModel> GetEnumerator()
        {
            return nameMap.Values.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }
    }
}

File: mapper/src/main/java/com/azure/autorest/models/clientmodel/ClientModels.java
package com.azure.autorest.models.clientmodel;

import java.util.*;

// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.


/** 
 The collection of all client models stored for inheritance lookup.
*/
//C# TO JAVA CONVERTER TODO TASK: The interface type was changed to the closest equivalent Java type, but the methods implemented will need adjustment:
public class ClientModels
{
	public static final ClientModels Instance = new ClientModels();

	private ClientModels()
	{
	}

	private final Map<String, ClientModel> nameMap = new HashMap<String, ClientModel>();
	private final Map<String, ArrayList<ClientModel>> derivedTypesMap = new HashMap<String, ArrayList<ClientModel>>();

	public final ClientModel GetModel(String modelName)
	{
		return nameMap.containsKey(modelName) ? nameMap.get(modelName) : null;
	}

	public final void AddModel(ClientModel model)
	{
		nameMap.put(model.getName(), model);

		ClientModel parentModel = model.getParentModel();
		if (parentModel != null)
		{
			ArrayList<ClientModel> derivedTypesList = GetDerivedTypeList(parentModel.getName());
			derivedTypesList.add(model);
		}
	}

	public final List<ClientModel> GetDerivedTypes(String parentModelName)
	{
		return GetDerivedTypeList(parentModelName);
	}

	private ArrayList<ClientModel> GetDerivedTypeList(String parentModelName)
	{
		if (!derivedTypesMap.containsKey(parentModelName))
		{
			derivedTypesMap.put(parentModelName, new ArrayList<ClientModel>());
		}
		return derivedTypesMap.get(parentModelName);
	}

	public final Iterator<ClientModel> iterator()
	{
		return nameMap.values().iterator();
	}

	public final Iterator GetEnumerator()
	{
		return GetEnumerator();
	}
}